File: permute.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <cmath>
#include "tree.h"

using namespace std;

vector<vector<int> > permute0(vector<int> &num);

vector<vector<int> > res;

void perm(vector<int> &num, int i) {
    if(num.size() == i){  
        res.push_back(num);  
    }
    for (int j = i; j < num.size(); ++j) {
        swap(num[i], num[j]);
        perm(num, i + 1);
        swap(num[i], num[j]);
    }
}

vector<vector<int> > permute(vector<int> &num) {  
    int N = num.size();  
    perm(num, 0);
    return res;
}

int main() {
    int a[] = {1, 2, 3, 4};
    vi b = buildVI(a, 4);
    vector<vector<int> > res = permute(b);
    printVVI(res);

    return 0;
}

// http://blog.csdn.net/tuantuanls/article/details/8717262
vector<vector<int> > permute0(vector<int> &num) {  
    int N = num.size();  
    vector<vector<int> > ret;  
    if(N == 1){  
        ret.push_back(num);  
        return ret;  
    }  
    vector<vector<int> > post;  
    vector<int> cur;  
    vector<int> tmp;  
    for(int i = 0; i < N; i++){  
        cur = num;  
        cur.erase(cur.begin()+i);  
        post = permute0(cur);  
        for(int j = 0; j < post.size(); j++){  
            tmp = post[j];  
            tmp.insert(tmp.begin(), num[i]);  
            ret.push_back(tmp);  
        }  
    }  
    return ret;  
}  

File: permute.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Collections;
import java.util.*;

public class permute {
    public static class Solution {

        public void removeElt( int [] arr, int remIndex ) {
            int numElts = arr.length - ( remIndex + 1 ) ;
            System.arraycopy( arr, remIndex + 1, arr, remIndex, numElts ) ;
        }

        // http://blog.csdn.net/tuantuanls/article/details/8717262
        public List<List<Integer>> permute(int[] num) {
            List<List<Integer>> list = new ArrayList<List<Integer>>();
            if (num.length == 1) {
                list.add(new ArrayList<Integer>(Arrays.asList(1)));
                return list;
            }

            List<List<Integer>> post = new ArrayList<List<Integer>>();
            List<Integer> curr = new ArrayList<Integer>();
            int [] temp;
            List<Integer> tmp = new ArrayList<Integer>();

            for (int i = 0; i < num.length; i++) {
                for(int x : num)
                    if (x != num[i])
                        curr.add(x);
                curr = Arrays.asList(num);
                post = permute(temp);
                for (int j = 0; j < post.size(); j++) {
                    tmp = post.get(j);
                    tmp.add(0, num[i]);
                    list.add(tmp);
                }
            }
            return list;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {1, 2, 3};
        List<List<Integer>> res = result.permute(a);
        
        System.out.println(res);
    }
}
